### PR TITLE
Restructure functionselect

### PR DESCRIPTION
**Pull request by light-and-salt**
Separate aggregates from Q Functions (Q Functions = NONE + BIN + aggregates - COUNT)

Separate timeUnits from T Functions (T Functions = NONE + timeUnits)

Tested in Voyager2 and PoleStar.

### CODE REVIEW

File: src/components/functionselect/functionselect.js [hunk 1]
```diff
@@ -137,14 +137,14 @@ angular.module('vlui')
           } else {
             // TODO: check supported type based on primitive data?
             if (isT) {
-              scope.func.list.aboveFold = timeUnits.aboveFold;
-              scope.func.list.belowFold = timeUnits.belowFold;
+              scope.func.list.aboveFold = _.slice(timeUnits.aboveFold);
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: I don't understand why you have to slice for this case

File: src/components/functionselect/functionselect.js [hunk 2]
```diff
@@ -137,14 +137,14 @@ angular.module('vlui')
...
+              scope.func.list.belowFold = _.slice(timeUnits.belowFold);
             }
             else if (isQ) {
-              scope.func.list.aboveFold = aggregates.aboveFold;
+              scope.func.list.aboveFold = _.slice(aggregates.aboveFold);
               // HACK
               scope.func.list.aboveFold.splice(1, 0, 'bin'); // support 'bin' for quantitative fields
-              scope.func.list.belowFold = aggregates.belowFold;
+              scope.func.list.belowFold = _.slice(aggregates.belowFold);
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: Why do you have to slice `belowFold`?

**Comment by kanitw** [comment]
> Fix a bug where bin is duplicated in scope.func.list.aboveFold.

I think the description is pretty vague.  I think you mean we set  `func.list.aboveFold = aggregates.aboveFold` and thus splicing `func.list.aboveFold` cause side effect to `aggregates.aboveFold`, which is probably bad.  But I still don't understand why is that bad?  Is it because then you'll have multiple `bin` in the list?

In any case, the `splice` thing is already a hack and this is another hack to fix the hack. 
I wonder if this would only create more and more problems?

I don't understand why you don't just have aboveFold and belowFold for Q and T, which is the right grouping rather than having to do this messy splice because the group by `aggregate` and `timeUnit` doesn't match what you want.

**Comment by light-and-salt** [comment]
@kanitw Re-structuring is done. Please review (again).

File: src/components/functionselect/functionselect.js [hunk 3]
```diff
@@ -44,12 +44,15 @@ angular.module('vlui')
             'secondsmilliseconds'
           ]
         };
-        timeUnits.all = timeUnits.aboveFold.concat(timeUnits.belowFold);
+        temporalFunctions.all = _.concat(temporalFunctions.aboveFold, temporalFunctions.belowFold);
 
-        // aggregates for Q
-        var aggregates = {
+        // timeUnits = T functions - undefined
+        var timeUnits = _.pull(temporalFunctions.all, undefined);
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: From `pull`'s documentation

> Unlike `_.without`, this method mutates array

I don't understand why you want to mutate temporalFunctions.all.

File: src/components/functionselect/functionselect.js [hunk 4]
```diff
@@ -62,24 +65,11 @@ angular.module('vlui')
...
-        function getAggregates(type) {
-          // HACK
-          // TODO: make this correct for temporal as well
-          if (type === 'quantitative' ){
-            return aggregates.all;
-          }
-          return [];
-        }
+        // aggregates = Q Functions + COUNT - BIN - undefined
+        var aggregates = _.pull(_.concat(quantitativeFunctions.all, [COUNT]),
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: From `pull`'s documentation

> Unlike `_.without`, this method mutates array

I don't understand why you want to mutate temporalFunctions.all.  

In general, you should avoid causing side-effect unless you have convincing reason to do so.

File: src/components/functionselect/functionselect.js [hunk 5]
```diff
@@ -44,12 +44,15 @@ angular.module('vlui')
             'secondsmilliseconds'
           ]
         };
-        timeUnits.all = timeUnits.aboveFold.concat(timeUnits.belowFold);
+        temporalFunctions.all = _.concat(temporalFunctions.aboveFold, temporalFunctions.belowFold);
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: In fact, `temporalFunctions.all` never get used.  I don't know why you need to declare it at all.  Why don't merge it into the logic below.

File: src/components/functionselect/functionselect.js [hunk 6]
```diff
@@ -62,24 +65,11 @@ angular.module('vlui')
             'variance', 'variancep'
           ] // hide COUNT for Q in the UI because we dedicate it to a special "# Count" field
         };
-        aggregates.all = aggregates.aboveFold.concat(aggregates.belowFold)
-          .concat([COUNT]); // COUNT is a valid aggregate
+        quantitativeFunctions.all = _.concat(quantitativeFunctions.aboveFold, quantitativeFunctions.belowFold);
```
Comment (kanitw) [inline src/components/functionselect/functionselect.js]: In fact, `quantitativeFunctions` never get used.  I don't know why you need to declare it at all.  Why don't merge it into the logic below.

File: src/components/functionselect/functionselect.js (hunk 3 above)
Comment (light-and-salt) [inline src/components/functionselect/functionselect.js]: Done

**Comment by light-and-salt** [comment]
@kanitw I've cleaned up the `.all` fields. Hopefully this is cleaner.

**Pull request merged**